File: go/say/say.go
package say

import "fmt"

var unitsString = map[int64]string{1: "one", 2: "two", 3: "three", 4: "four", 5: "five", 6: "six", 7: "seven", 8: "eight", 9: "nine"}
var tenToNineteenString = map[int64]string{10: "ten", 11: "eleven", 12: "twelve", 13: "thirteen", 14: "fourteen", 15: "fifteen", 16: "sixteen", 17: "seventeen", 18: "eighteen", 19: "nineteen"}
var tensString = map[int64]string{2: "twenty", 3: "thirty", 4: "forty", 5: "fifty", 6: "sixty", 7: "seventy", 8: "eighty", 9: "ninety"}

func Say(n int64) (string, bool) {
	if n > 999999999999 || n < 0 {
		return "", false
	}
	numberString := ""
	billions := n / 1000000000
	millions := (n % 1000000000) / 1000000
	thousands := (n % 1000000) / 1000
	units := n % 1000
	if billions > 0 {
		numberString = fmt.Sprintf("%s billion", stringSmallNumbers(billions))
		if millions > 0 || thousands > 0 || units > 0 {
			numberString = fmt.Sprintf("%s ", numberString)
		}
	}
	if millions > 0 {
		numberString = fmt.Sprintf("%s%s million", numberString, stringSmallNumbers(millions))
		if thousands > 0 || units > 0 {
			numberString = fmt.Sprintf("%s ", numberString)
		}
	}
	if thousands > 0 {
		numberString = fmt.Sprintf("%s%s thousand", numberString, stringSmallNumbers(thousands))
		if units > 0 {
			numberString = fmt.Sprintf("%s ", numberString)
		}
	}
	if units > 0 {
		numberString = fmt.Sprintf("%s%s", numberString, stringSmallNumbers(units))
	} else if billions == 0 && millions == 0 && thousands == 0 {
		numberString = "zero"
	}
	return numberString, true
}

func stringSmallNumbers(n int64) string {
	numberString := ""
	hundreds := n / 100
	tens := (n % 100) / 10
	units := n % 10

	if hundreds > 0 {
		numberString = fmt.Sprintf("%s hundred", unitsString[hundreds])
		if tens > 0 || units > 0 {
			numberString = fmt.Sprintf("%s ", numberString)
		}
	}
	if tens > 1 {
		numberString = fmt.Sprintf("%s%s", numberString, tensString[tens])
		if units := (n % 10); units >= 1 {
			numberString = fmt.Sprintf("%s-%s", numberString, unitsString[units])
		}
	} else if tens == 1 {
		numberString = fmt.Sprintf("%s%s", numberString, tenToNineteenString[n % 100])
	} else if units != 0 {
		numberString = fmt.Sprintf("%s%s", numberString, unitsString[units])
	}
	return numberString
}

File: elixir/say/lib/say.ex
defmodule Say do
  @digits %{0 => "", 1 => "one", 2 => "two", 3 => "three", 4 => "four", 5 => "five", 6 => "six", 7 => "seven", 8 => "eight", 9 => "nine"}
  @tens %{0 => "", 2 => "twenty", 3 => "thirty", 4 => "forty", 5 => "fifty", 6 => "sixty", 7 => "seventy", 8 => "eighty", 9 => "ninety"}
  @teens %{0 => "ten", 1 => "eleven", 2 => "twelve", 3 => "thirteen", 4 => "fourteen", 5 => "fifteen", 6 => "sixteen", 7 => "seventeen", 8 => "eighteen", 9 => "nineteen"}

  @doc """
  Translate a positive integer into English.
  """
  @spec in_english(integer) :: {atom, String.t()}
  def in_english(number) when number < 0 or number > 999999999999, do: {:error, "number is out of range"}
  def in_english(0), do: {:ok, "zero"}
  def in_english(number) do
    {:ok, number
    |> Integer.digits
    |> Enum.reverse
    |> Enum.chunk_every(3)
    |> Enum.map(&Enum.reverse/1)
    |> Enum.zip(["", " thousand", " million", " billion"])
    |> Enum.reverse
    |> Enum.map(&construct_words/1)
    |> Enum.filter(&(&1 != ""))
    |> Enum.join(" ")}
  end

  defp construct_words({num, scale}) do
    if (words = construct_phrase(num)) == "", do: "", else: words <> scale
  end

  defp construct_phrase([0, tens, units]), do: tens_and_units([tens, units])
  defp construct_phrase([hundreds, tens, units]), do: "#{@digits[hundreds]} hundred" <> if (word = tens_and_units([tens, units])) != "", do: " #{word}", else: ""
  defp construct_phrase(num), do: tens_and_units(num)

  defp tens_and_units([1, units]), do: @teens[units]
  defp tens_and_units([0, units]), do: @digits[units]
  defp tens_and_units([tens, 0]), do: @tens[tens]
  defp tens_and_units([tens, units]), do: "#{@tens[tens]}-" <> @digits[units]
  defp tens_and_units([units]), do: @digits[units]
end
